distribution/tools/plugin-cli/src/main/java/org/elasticsearch/plugins/PluginsManifest.java

```diff
@@ -0,0 +1,151 @@
+            throw new RuntimeException("Cannot have null or empty plugin IDs in: " + manifestPath);
+        }
+
+        final Set<String> uniquePluginIds = new HashSet<>();
+        for (final PluginDescriptor plugin : plugins) {
+            if (uniquePluginIds.add(plugin.getId()) == false) {
+                throw new UserException(ExitCodes.USAGE, "Duplicate plugin ID [" + plugin.getId() + "] found in: " + manifestPath);
+            }
+        }
+
+        for (PluginDescriptor plugin : this.plugins) {
+            if (InstallPluginAction.OFFICIAL_PLUGINS.contains(plugin.getId()) == false && plugin.getUrl() == null) {
+                throw new UserException(
+                    ExitCodes.CONFIG,
+                    "Must specify URL for non-official plugin [" + plugin.getId() + "] in " + manifestPath
+                );
+            }
+        }
+
+        if (this.proxy != null) {
+            final String[] parts = this.proxy.split(":");
+            if (parts.length != 2) {
+                throw new UserException(ExitCodes.CONFIG, "Malformed [proxy], expected [host:port] in: " + manifestPath);
+            }
+
+            if (ProxyUtils.validateProxy(parts[0], parts[1]) == false) {
+                throw new UserException(ExitCodes.CONFIG, "Malformed [proxy], expected [host:port] in: " + manifestPath);
+            }
+        }
+
+        for (PluginDescriptor p : plugins) {
+            if (p.getUrl() != null) {
+                try {
+                    new URL(p.getUrl());
+                } catch (MalformedURLException e) {
+                    throw new UserException(ExitCodes.CONFIG, "Malformed URL for plugin [" + p.getId() + "]");
+                }
+            }
+        }
+    }
+
+    /**
+     * Constructs a {@link PluginsManifest} instance from the specified YAML file, and validates the contents.
+     * @param env the environment to use in order to locate the config file.
+     * @return a validated manifest
+     * @throws UserException if there is a problem finding, parsing or validating the file
+     */
+    public static PluginsManifest parseManifest(Environment env) throws UserException {
+        final Path manifestPath = env.configFile().resolve("elasticsearch-plugins.yml");
+        if (Files.exists(manifestPath) == false) {
+            throw new UserException(ExitCodes.CONFIG, "Plugin manifest file missing: " + manifestPath);
+        }
+
+        final YAMLFactory yamlFactory = new YAMLFactory();
+        final ObjectMapper mapper = new ObjectMapper(yamlFactory);
+
+        PluginsManifest pluginsManifest;
+        try {
+            byte[] manifestBytes = Files.readAllBytes(manifestPath);
+            pluginsManifest = mapper.readValue(manifestBytes, PluginsManifest.class);
+        } catch (IOException e) {
+            throw new UserException(ExitCodes.CONFIG, "Cannot parse plugin manifest file [" + manifestPath + "]: " + e.getMessage());
```

Comment:
We should also include the original exception as a cause so we have the full stacktrace.

Reply:
Added the original exception.